src/v/iceberg/values.cc

```diff
@@ -0,0 +1,387 @@
+    }
+    size_t operator()(const timestamptz_value& v) const {
+        return std::hash<int64_t>()(v.val);
+    }
+    size_t operator()(const string_value& v) const {
+        return std::hash<iobuf>()(v.val);
+    }
+    size_t operator()(const uuid_value& v) const {
+        return absl::Hash<uuid_t>()(v.val);
+    }
+    size_t operator()(const fixed_value& v) const {
+        return std::hash<iobuf>()(v.val);
+    }
+    size_t operator()(const binary_value& v) const {
+        return std::hash<iobuf>()(v.val);
+    }
+    size_t operator()(const decimal_value& v) const {
+        return absl::Hash<absl::int128>()(v.val);
+    }
+};
+
+struct hashing_visitor {
+    size_t operator()(const primitive_value& v) const {
+        return std::visit(primitive_hashing_visitor{}, v);
+    }
+
+    size_t operator()(const std::unique_ptr<struct_value>& v) const {
+        if (!v) {
+            return 0;
+        }
+        return value_hash(*v);
+    }
+    size_t operator()(const std::unique_ptr<list_value>& v) const {
+        if (!v) {
+            return 0;
+        }
+        size_t h = 0;
+        for (const auto& e : v->elements) {
+            if (!e) {
+                continue;
+            }
+            boost::hash_combine(h, std::hash<value>()(*e));
+        }
+        return h;
+    }
+    size_t operator()(const std::unique_ptr<map_value>& v) const {
+        if (!v) {
+            return 0;
+        }
+        size_t h = 0;
+        for (const auto& kv : v->kvs) {
+            boost::hash_combine(h, std::hash<value>()(kv.key));
+            if (kv.val) {
+                boost::hash_combine(h, std::hash<value>()(*kv.val));
+            }
+        }
+        return h;
+    }
+};
+
+void ostream_val_ptr(std::ostream& o, const std::optional<value>& v) {
+    if (v) {
+        o << *v;
+        return;
+    }
+    o << "none";
+}
+
+} // namespace
+
+struct primitive_value_comparison_visitor {
+    template<typename T, typename U>
+    bool operator()(const T&, const U&) const {
+        static_assert(!std::is_same<T, U>::value);
+        return false;
+    }
+    template<typename T>
+    requires requires(T t) { t.val; }
+    bool operator()(const T& lhs, const T& rhs) const {
+        return lhs.val == rhs.val;
+    }
+};
+
+bool operator==(const primitive_value& lhs, const primitive_value& rhs) {
+    return std::visit(primitive_value_comparison_visitor{}, lhs, rhs);
+}
+
+bool operator==(const struct_value& lhs, const struct_value& rhs) {
+    if (lhs.fields.size() != rhs.fields.size()) {
+        return false;
+    }
+    for (size_t i = 0; i < lhs.fields.size(); i++) {
+        auto has_lhs = lhs.fields[i] != std::nullopt;
+        auto has_rhs = rhs.fields[i] != std::nullopt;
+        if (has_lhs != has_rhs) {
+            return false;
+        }
+        if (!has_lhs) {
+            // Both are null.
+            continue;
+        }
+        if (*lhs.fields[i] != *rhs.fields[i]) {
```

Comment:
Are the fields in a struct guaranteed to be in the same order? In protobuf they're not. E.g. are these two structs equal?

```
a = {"number": 5, "message": "hello"}
b = {"message": "hello", "number": 5}
```

Reply:
In this definition they're not equal. Those kinds of semantics are things I don't think are called out explicitly in the Iceberg spec (esp given they're serialization dependent). Opted to leave this value representation flexible by leaving that distinction to callers when they construct values (e.g. maybe a protobuf translator would be required to order the kvs).